### ISSUE TITLE
Configuration data layout

### ISSUE THREAD

**Issue by jgharston** [kgl2001/IntegraB-OS#14]
Is the configuration data layout exposed via OSBYTE 161/162 intended to be the same as the Master and other platforms:
https://beebwiki.mdfs.net/CMOS_configuration_RAM_allocation

**Comment by kgl2001** [comment]
Interesting. The IntegraB registers do seem to mirror the Master registers, where appropriate. It also explains why there are gaps in the IntegraB register map. We might need to have a rethink about what we've done with registers &5 and &11 in the v1.26 release.
```
rtcUserBase = &0E
;RTC User Registers (add &0E to get real RTC user register - registers &00-&0D are for RTC clock registers)
;The hex values shown after the register numbers are the defaults, where these are simple
;enough to include. Defaults may vary with IBOS version and the full details can be seen in the
;code at UserRegDefaultTable.
;Register &00
;Register &01
;Register &02
;Register &03
;Register &04
;Register &05 -               <v1.26: 0-3: FILE, 4-7: LANG / >=v1.26: 0-3: non-tube-LANG, 4-7: tube-LANG
;Register &06 - &FF:	*INSERT status for ROMS &0F to &08. Default: &FF (All 8 ROMS enabled)
;Register &07 - &FF:	*INSERT status for ROMS &07 to &00. Default: &FF (All 8 ROMS enabled)
;Register &08
;Register &09
;Register &0A - &17/&E7:	0-2: MODE / 3: SHADOW / 4: TV Interlace / 5-7: TV screen shift
;Register &0B - &23:	0-2: FDRIVE / 3-5: CAPS
;Register &0C - &19:	0-7: Keyboard Delay
;Register &0D - &05:	0-7: Keyboard Repeat
;Register &0E - &0A:	0-7: Printer Ignore
;Register &0F - &2D:	0: Tube / 2-4: BAUD / 5-7: Printer
;Register &10 - &A0:	0: File system disc/net flag / 4: Boot / 5-7: Data
;Register &11 - &FF:          <v1.26: unused / >=v1.26: 0-3: FILE, 4-7: spare
```

**Comment by jgharston** [comment]
While internally it doesn't matter what is stored where, it is important that to a user calling OSBYTE 161/162 they get the correct results as used elsewhere, eg Master/Compact, Master Extension ROM, Plus1 Support, etc.

**Comment by ZornsLemma** [comment]
This is interesting, I hadn't thought about Master etc compatibility here.

Of course, since this "dual-LANG" concept added in v1.26 is IBOS-specific, any general-purpose code which is reading/writing CMOS directly isn't going to understand this and at best will only be able to modify one of the LANG options.

Looking at https://beebwiki.mdfs.net/CMOS_configuration_RAM_allocation, our use of &11 (17) is going to clash with ANFS, but in practice that's probably fine, because the B/B+ version of ANFS isn't going to use CMOS RAM and the Master version isn't going to run on a B/B+.

I don't know if there's a perfect solution, but my initial thoughts are to tweak IBOS so we store FILE in the low bits of &05, which is where it's always lived and where the Master has it too, and use &11 for LANG and accept that any general purpose code which is trying (I'm not sure why) to read/write the *CONFIGURE LANG settings isn't going to work on IBOS. In other words, make userRegFile=&05 and userRegLang=&11 in v1.27, the opposite of v1.26.

We could also perhaps permanently store the IBOS bank number in the high bits of &05, so it looks like the IBOS NLE is the current language to any code which does try to read it.

Changing &11 to a different address would also be easy if either of you has any suggestions. We could use a much higher register; IIRC the Integra-B has everything battery backed, but *CSAVE/*CLOAD save/load 128 bytes so we might say that for "config" purposes there are 128 bytes of CMOS RAM, and we could stop using &11 and switch to a register which doesn't even exist on the Master 128 to keep it safely out of the way. 

I am willing to tweak the code to change the storage beyond just rearranging bytes. This is probably going to make it a bit larger though. We could perhaps keep a write-only copy of (say) the non-tube LANG setting in the high bits of &05, just for the benefit of any code which tries to read it, although I'm not sure it's worth it.

**Comment by ZornsLemma** [comment]
I just experimented in b-em on a Master and it looks like there if you *CONFIGURE LANG to be HIBASIC and disable the tube, it drops to a * prompt. If we're really interested in Master compatibility, we could drop the dual-LANG support added in v1.26 and just retain the "fall back to NLE if it's a HI language and tube isn't present" code. I can't help thinking the ability to boot to BASIC or HIBASIC automatically is nifty though.

**Comment by ZornsLemma** [comment]
While I think it's obviously sensible to minimise the special case logic in *FX161/162, as Jonathan points out, we can do things in there to hide the differences in the underlying layout. So maybe *FX161 reads return (say) the non-tube LANG setting when reading the high bits of &05, and *FX162 writes to the high bits of &05 set both the tube and non-tube LANG to the same bank (just as *CONFIGURE LANG n does).

Doing something like this is definitely nicer, but I wonder if it's worth the code it would take to implement it. We should almost certainly move FILE back to &05 where it's always been and is on the Master, but I'm not so sure about trying to support code reading/writing the LANG setting.

**Comment by kgl2001** [comment]
So, my thoughts on this are...

- There are 50 registers available in the CDP6818 (I assume the Master HD14618 is the same).
- The beebwiki makes reference to all 50 registers for the Master, although some are reserved, and could be used.
- Retain the current tube LANG and non-tube LANG configuration.
- Revert FILE back to to register &05, and use reserved register &15 (21) for tube LANG.
- Don't do anything fancy with *FX161/162. Just read / write directly from / to the register.

**Comment by ZornsLemma** [comment]
That seems reasonable in general - just to clarify, are you suggesting we store both LANG values in the same register (as we do now, which is the easy/simple code option), or that non-tube LANG lives in &5 and the tube LANG lives in &15 (a bit fiddlier in terms of the code)?

You are (of course) right about the CDP6818 itself having 50 registers/bytes of general purpose RAM, but from a software perspective, {Read,Write}UserReg use some of the 12K private RAM to extend this to 128 bytes and *CSAVE/*CLOAD save/load all 128 bytes. So I *think* - unless I'm missing something - for our purposes here the actual underlying CDP6818 is abstracted away. It looks to me (judging from the comment at around line 151 of IBOS.asm) as though there are lots of these extra pseudo-registers free for us to use if we choose.

That said, I don't see a big problem with using &15, as long as no one else needs it. It just feels as though this is a (not very, but slightly) scarce community resource we would be allocating from, when we can do more-or-less as we please with the (pseudo) registers >=50.

**Comment by kgl2001** [comment]
My initial thoughts were to split LANG over the two registers, but if that comes at a byte cost (even relatively low) in the ROM, then perhaps not.

Regarding the 50 register limit you are, of course, quite correct. IBOS does indeed extend this into the Private RAM area. So perhaps these Private registers should be used for IntegraB specific configuration settings.  However, I note from [this](https://beebwiki.mdfs.net/OSBYTE_%26A1) Wiki entry that the Master Compact uses a 128 byte EEPROM, so arguably this Private space may also have been allocated for other uses. I suspect this is highly unlikely, though, and I really don't think we're going to run into any compatibility issues, whichever way we go here!

**Comment by ZornsLemma** [comment]
Thanks Ken! Looking over the code and thinking as I go, to split LANG over two registers:
- ~~FullReset would need to special case an extra register, perhaps costing four bytes.~~ Edit: No, this is a wash. We currently special case two registers and we'd continue to special case two registers.
- SelectConfiguredFilingSystemAndLanguage would need to set X to the appropriate register based on tube presence or absence before calling ReadUserReg. On the flip side, if we used the high nybble of both registers, we wouldn't need the conditional right shift. So this is probably a wash, maybe a few bytes longer.
- ConfLang would need to stop using {Get,Set}ConfigValue, because those are oriented around accessing a specific byte for any given config option. We'd need to use {Read,Write}UserReg directly to read/write the relevant nybbles of each of the two separate registers. This is probably the major code size hit from splitting the two up. Without writing a sketch of the code it's hard to give an over precise estimate, but we're probably looking at something like 10-20 extra bytes here.

I agree we're unlikely to run into big compatibility issues with the choice of our extra byte of "CMOS" to replace &11. Allocating it in the Compact-only range does feel slightly safer though, and leaves &15 free for things that need to be Master 128 compatible in the future. Just to be concrete, shall I propose "register" &7E, which is right at the end of the free space?

Edit: Not terribly plausible, but in principle someone might feel motivated to hack the Master OS along the lines of [OS 1.26](https://github.com/regregex/OS126) for the BBC B and want to copy IBOS's dual-LANG idea, so using a Master 128-compatible CMOS location might be nice. But that's already unlikely, and the idea of anyone writing software that wants to tweak this setting via *FX161/162 feels like an extra layer of improbability. So I'd still suggest &7E, but I thought this was worth mentioning as I find it an amusing idea.

**Comment by kgl2001** [comment]
Register &7E sounds like an appropriate choice, but picking up on your comment about implementing dual-LANG in the Master OS, this should definitely be a consideration. So using register &15 may not be such a wild idea. Maybe one for @jgharston to comment on. That extra 10-20bytes to split out LANG into two registers is a hard pill to swallow, though!

**Comment by ZornsLemma** [comment]
I definitely lean towards not splitting LANG up, in the absence of a good example of a non-IBOS-specific user program that is going to want to read/write LANG programmatically, and bearing in mind that such a program will only ever see part of the dual-LANG picture even if we do go out of our way to support it by keeping one of the two settings at &05 or making it appear at &05 via *FX161/162. And if it's a "new" program written by a retro-enthusiast - perhaps some cool bells-and-whistles graphical configuration tool - they might well be willing to accomodate IBOS's dual-LANG explicitly.

Edit: And as far as writing LANG programmatically goes, if the program issues *CONFIGURE LANG n it will work fine; it's only writes via *FX162 that won't. So it's programmatic reads of LANG which are the bigger aspect of this, I think.

One very naive implementation of dual-LANG for the Master would be a utility ROM which implements a language entry and is set as the OS 3.20 default language via register &05, and on its language entry examines our alternate dual-LANG byte and then enters the appropriate language. This feels clunky and if anyone cared I'm sure there'd be a better way, but this *would* perhaps help justify leaving &05 alone as far as LANG goes and putting both LANG settings together in a different register <50.

Edit: I could imagine a user program wanting to load/save the CMOS RAM programmatically as a whole, for switching between different ways of using the machine. Obviously IBOS has *CSAVE/*CLOAD for this, but someone could potentially write a general purpose program to do this, and they'd naturally use *FX161/162 to read/write the entire contents of CMOS, but without caring about what individual bytes mean. This example makes me think that putting the dual-LANG config in the Master 128-compatible range *or* supporting reads from "register" &FF to determine the CMOS size (so this hypothetical program will save everything correctly, if it's well-written) wouldn't be a bad idea.

**Comment by kgl2001** [comment]
So, are you suggesting that we store both LANG settings in the same register, and possibly in a Master register other than &05 (eg &15)? And modify *FX161/162 so that read and write of the &05 LANG parameter is remapped to this other register? If so, would keeping a copy of non-tube LANG in register &05 help in any way? Or would that just complicate matters further???

**Comment by ZornsLemma** [comment]
I don't feel very strongly about most of this - except not being keen to add extra code to support splitting LANG up into two registers (wherever they are), and (which I think is completely uncontroversial at this point) thinking we should moving FILE back to &05. 

That said, yes, I guess I am making a (weak) argument for storing both LANG settings in the same non-&05 register in the Master-compatible range, such as &15. I'm not particularly suggesting we should do anything clever with *FX161/162; I said a way above it would be "nicer", but I don't really think it's worth the code it would take to do this, I was more throwing out options/ideas.

I say the argument is weak because we'd need someone to a) care about dual-LANG on the Master 128 and implement it *and* b) write a program that uses *FX161/162 to manipulate the dual-LANG settings (instead of just using *CONFIGURE LANG a,b to set it and not needing to read it except via *STATUS) *and* c) for it to be desirable that that program can be used without any tweaks on an IBOS 1.27+ Integra-B as well. All of these are possible, and the downsides to keeping this option open by using register &15 are slim, given in the 35-ish years since the M128 came out, only one byte of the reserved space has been allocated (assuming BeebWiki is accurate, which it probably is) so us taking a byte of it here isn't really likely to cause a shortage of reserved space. But all of these happening feels implausible, so it wouldn't be hard to argue that we should just use &7E and keep our hands off that "communal" space because of the low probability of a+b+c. There's something to be said for both points of view and it's ultimately a judgement call, of course.

If we keep a copy of the non-tube LANG in register &05 that would require a bit of extra code (but as much as splitting LANG completely) and in return any general non-IBOS-specific code that attempts to programmatically read the current LANG from &05 would see a semi-sensible value. I'm not terribly keen to do this, as I can't really picture a program wanting to read the current LANG (as opposed to just "read the whole of CMOS, without interpreting individual bytes") in the first place.

Edit: Apologies for perhaps spamming this issue a little bit with various thoughts, which is maybe annoying and/or confusing. I'm happy to implement this whatever we decide, including splitting up the dual LANG setting across two registers. My verbosity on this is in part because it seems to me to be a bit of a fuzzy issue of judgement rather than something that seems to have a clear right/wrong.

Edit 2: TL;DR: I mostly agree with your post above starting "So, my thoughts" with the bullet-pointed list, except that:
- I favour (but by no means insist on) using a single register for both non-tube and tube LANG values
- I'm on the fence about which register we should store the dual LANG values in, with the two main candidates being &11 and &7E.

**Comment by kgl2001** [comment]
I'm ok with single register for both non-tube and tube LANG values. Does the upper nybble of register &05 just become spare, or do we store the IBOS bank number in there as you suggested earlier?

I'm also not overly excited about which register we use for the dual LANG settings either. If I had the casting vote, I'd pick &11. I'd still be interested to hear @jgharston thoughts on this, though.

**Comment by jgharston** [comment]
Any ROM can use ROMNUMBER+30 to store its own settings. This would obviously be an option to the IBOS ROM itself.

**Comment by jgharston** [comment]
I've made a table of IBOS configuration data vs common data used by other systems (not just Master, things like Beebug's Master Extension ROM, Plus 1, other people's Master-stuff-for-Beeb, etc.)

Locations 6 and 7 are the wrong way around. bit 0 of location 6 is the frugal bit for ROM 0, bit 7 of location 7 is the frugral bit for ROM 15. Code everywhere else can do A%=161:X%=6:U%=USR(&FFF4):IF U% AND &10000 THEN P."ROM 0 unplugged" (had to quickly check this on a real Master to check the documentation! - *FX162,6,254 should leave ROM 0 unplugged)

While ANFS on the BBC doesn't use the config. settings, many net tools do. Bit 7 of location &11 indicates if a utility should display a startup message. Many programs do: LDA #161:LDX #17:LDY #0:JSR OSBYTE:TYA:BPL nomessage. If there is no OSBYTE 161 support, Y is returned unchanged - so preloading it with 0 means it returns either with 0, or with the actual value.

IBOS puts the date century in location 51, everywhere else it is in b6-b7 of location 18.
IBOS puts NFS/DFS in bit 0 of location 16, everywhere else uses this to store *Configure Shadow
IBOS can use its own ROMNUM+30 location, just as any ROM can.

With this in mind, I'd recommend:
Put locations 6 and 7 the other way around to match everybody else
Move Century to location 18 b7-b6. Might be simpl;est to read it on startup and expand to 19+nn and put it somewhere locally for the rest of the code to use.
Return location 5 to b0-b3=File, b4-b7=Lang
That leaves 9 bits of information to allocate. ROMNUM+30 gives 8 bits, that could be:
3 bits for OSMODE
1 bit for SHX
4 bits for HILANG
That leaves us needing 1 bit for DFS/NFS. The Beebug MER uses bit 4 of location 9 for this. It is allocated as "Telecoms software", but I haven't tracked down anything other that uses it, other than DFS/NFS.
On top of this, there are two bits in location 8 that are unallocated.

**Comment by jgharston** [comment]
Replying to the "user program accessing config data with two LANG settings" points above:
Yes, you could expect or require a program to write to the config data with *CONFIGURE LANG, but the only way of reading is with OSBYTE 161. A program such as the Control Panel demo reads and writes with OSBYTE, and there are a handful of config editors that do similar things, and programs generally expect to be able to do, eg: A%=161:X%=5:Y%=((USR&FFF4)AND&FF00)DIV65536:P."File ";Y%AND15;"  Lang: ";Y%DIV16.

**Comment by jgharston** [comment]
I'm a bit lost in the code, but having two LANG nybbles would need code something like this whereever the current code is:

LDX #userRegLang
BIT tubePresent:BPL here
JSR GetMyROMByte ; LDA ROMID,CLC,ADC #30,TAX,RTS
.here
JSR ReadUserReg
JSR LSRtimes4
STA osbyteCurrentLanguage

Essentially 8 extra bytes, seeing as GetMyROMByte should also be needed for other code.

**Comment by ZornsLemma** [comment]
Thanks Jonathan! This seems reasonable on the whole, although I am a bit concerned about the user experience of upgrading/downgrading IBOS if we shuffle a lot of stuff round in the CMOS RAM. Putting that to one side, it would obviously be nice if IBOS did do things compatibly with other systems.

Since IBOS does have 128 bytes of "CMOS", I think we'd be better off using a byte of the "extra" RAM (say &7E) instead of the byte allocated to the ROM in IBOS's bank at 30+X. This would (although it's a bit of a corner case) help make it easier to test alternate versions of IBOS in a different bank, and it would offer some small code size savings, e.g. your code fragment above would become

LDX #userRegLang
BIT tubePresent:BPL here
LDX #&7E
.here
JSR ReadUserReg 
...

I think the LANG nybble is only currently read explicitly in one place (at SelectConfiguredFilingSystemAndLanguage), the other is at the ConfLang label where we use {Get,Set}ConfigValue to read/write it - it is selected implicitly via an index in the Y register. There would be some extra code overhead in making ConfLang capable of reading two separate bytes for *STATUS output and updating part of two separate bytes when handling *CONFIGURE LANG.

I am probably being a bit over-frugal in terms of keeping the code as small as possible and it does look like there's value in keeping the non-tube LANG setting for IBOS in the "standard" place. Of course any generic tool that tries to set LANG using *FX162 isn't going to work perfectly on IBOS even if we have kept non-tube LANG in &05's high nybble, as it won't know there's a separate HILANG, but I suppose it is nice if it at least sort of works. I see you've identified some additional code savings already (Ken sent me a copy of the notes you sent him, and I see you've raised an issue with at least one big code saving) on top of the ones I've marked in the code with SQUASH:, so while it feels almost dangerously extravagant, we probably can afford this here if we want. :-)

Taking off my "keep the code size down" hat, I did wonder about making *FX162 special-case writes to &05 so the LANG nybble always gets duplicated into the HILANG nybble as well. Given that behaviour, a generic config-editing tool would set both LANG nybbles automatically, which seems like the best case for a tool that isn't aware of IBOS's dual LANG support. However, on reflection I don't like that idea even if code space isn't a concern, because it would make it hard to do a simple save/restore of all the config data via *FX161/162, as you'd need to know (which generic code won't) that you must write &05 *before* you write whichever register has HILANG in in order to work around this special case write.

Whether we use IBOS's allocated ROM byte in CMOS at 30+X or &7E to store HILANG, the somewhat hypothetical case of being compatible in terms of CMOS use with a Master implementation of "dual LANG" isn't possible, since on the Master there's no &7E byte and no IBOS ROM with an associated allocated byte. I don't think this is a huge deal but I did want to point it out given the discussion I was having with Ken yesterday.

**Comment by kgl2001** [comment]
> I am a bit concerned about the user experience of upgrading/downgrading IBOS if we shuffle a lot of stuff round in the CMOS RAM.

I don't think this is really a big issue, as long as we're not changing the config around at every revision. Perhaps we consider stepping the revision up to a more major release number (eg 1.30) if we do make any major changes to the config table layout for compatibility reasons?

More generally, I'm not aware of any software compatibility issues arising from the Master config table being slightly different to the IBOS config table, so arguably, why change something that's been working ok for 30ish years? I'm conscious we could break some software that is already IBOS config table aware if we do make changes. I'm thinking specifically about the Level 9 graphic adventures here, that specifically take advantage of the IntegraB shadow and sideways RAM. I suspect it's not poking around the CMOS config to determine if shadow or sideways RAM is available, so it's unlikely to be an issue, but definitely worth checking we haven't broken it after making any changes. 

Bottom line is that I think the risk of breaking something as a result of trying to align the IBOS table with the Master table is pretty small, and I do like the thought of being consistent, so ultimately I would be happy to align the IBOS config table more closely with the Master table, as long as it doesn't cost too much in ROM bytes.

However, it still leaves open the question of what we should do with non-tube LANG and tube LANG. It sounds like the 'compatible' thing to do would be to split it over two registers; keeping non-tube LANG at &05, and relocating tube LANG to a different location. Would we be able to minimise the byte cost by storing the two LANG values in the same nybble of the respective registers?

**Comment by jgharston** [comment]
Having the two LANG settings in two bytes, both in b7-b4 make it simpler, as you're not having different bit manipulation depending on which byte you read. You get something like:
LDX #5:BIT tube:BPL xxxx:LDX #&7E
.xxxx
JSR readconfigbyte:JSR LSRA4:STA osbyteCurrentLanguage

I'm seeing the attraction to having IBOS-specific stuff up at the top end of the config settings. "Officially" there are 50 bytes, so putting stuff at the far end of 128 and working down keeps things away. Any *CMOSSAVE command is expected to use the appropriate call to ask how many bytes exist and save them all, so having more than 50 is no issue to correctly-written programs.

**Comment by jgharston** [comment]
The main things to align the layout with other configuation implementations is just:
* location 5 is lang/file
* swap 6/7
* don't touch 16 bit 0 (unless you implement *Configure Shadow)
* don't touch 17 bit 7 (utilities will display extended help if set)
* put the century in 18 bit 7-6, or ignore b7 and just implement 19/20 via bit 6 - who's going to be alive in 2100? :)

Compared to IBOS 1.20 all these are probably code neutral, other than moving the Century*. Optimisations and cleaing up the code is well on the way to releasing space for improvements and extensions.

* six bytes? LDX #18:JSR readconfig:ASL A:ASL A:LDA #19:ADC #0
